apisix/balancer/ewma.lua

```diff
@@ -0,0 +1,256 @@
+-- swap the value at position i with the value at position r
+local function shuffle_peers(peers, k)
+    for i = 1, k do
+        local rand_index = math.random(i, #peers)
+        peers[i], peers[rand_index] = peers[rand_index], peers[i]
+    end
+    -- peers[1 .. k] will now contain a randomly selected k from #peers
+end
+
+
+local function pick_and_score(peers, k)
+    shuffle_peers(peers, k)
+    local lowest_score_index = 1
+    local lowest_score = score(peers[lowest_score_index])
+    for i = 2, k do
+        local new_score = score(peers[i])
+        if new_score < lowest_score then
+            lowest_score_index, lowest_score = i, new_score
+        end
+    end
+    return peers[lowest_score_index], lowest_score
+end
+
+
+-- slow_start_ewma is something we use to avoid sending too many requests
+-- to the newly introduced endpoints. We currently use average ewma values
+-- of existing endpoints.
+local function calculate_slow_start_ewma(self)
+    local total_ewma = 0
+    local endpoints_count = 0
+
+    for _, endpoint in pairs(self.peers) do
+        local endpoint_string = endpoint.address .. ":" .. endpoint.port
+        local ewma = shm_ewma:get(endpoint_string)
+
+        if ewma then
+            endpoints_count = endpoints_count + 1
+            total_ewma = total_ewma + ewma
+        end
+    end
+
+    if endpoints_count == 0 then
+        ngx.log(ngx.INFO, "no ewma value exists for the endpoints")
+        return nil
+    end
+
+    return total_ewma / endpoints_count
+end
+
+
+local function _trans_format(t1)
+    -- trans
+    --{"1.2.3.4:80":100,"5.6.7.8:8080":100}
+    -- into
+    -- [{"address":"1.2.3.4","port":"80"},{"address":"5.6.7.8","port":"8080"}]
+    local t2 = {}
+    local addr, port, err
+    local k,_
+
+    for k,_ in pairs(t1) do
+        addr, port, err = core.utils.parse_addr(k)
+        if not err then
+            t2[#t2+1] = {address = addr, port = tostring(port)}
+        else
+            core.log.error('parse_addr error: ',k,err)
+        end
+    end
+    return next(t2) and t2 or nil
+end
+
+
+local function _ewma_find(up_nodes)
+    local peers
+    local endpoint, ewma_score
+    local err
+
+    if not ewma_lock then
+        ewma_lock, ewma_lock_err = resty_lock:new("balancer_ewma_locks", {timeout = 0, exptime = 0.1})
+    end
+    if not ewma_lock then
+        return nil, ewma_lock_err
+    end
+    peers = _trans_format(up_nodes)
+    if not peers then
+        err = 'up_nodes error'
+        return nil, err
+    end
+    endpoint, ewma_score = peers[1]
+    if #peers > 1 then
+        local k = (#peers < PICK_SET_SIZE) and #peers or PICK_SET_SIZE
+        endpoint, ewma_score = pick_and_score(peers, k)
+    end
+    --ngx.var.balancer_ewma_score = ewma_score
+
+    return endpoint.address .. ":" .. endpoint.port
+end
+
+
+local function _ewma_after_balance()
+    local response_time = tonumber(ngx.var.upstream_response_time) or 0
```

Comment:
fetch var from `ngx.var` is bad performance.

in APISIX, we should always use `ctx.var.****` which is better performance.